HDDS-8050. REST Endpoint for S3 secret manipulation

ivanzlenko opened the pull request:
## What changes were proposed in this pull request?
Created new endpoint for managing S3 secrets. 
Endpoint can be accessed using path /secret. 
Available commands: 
* generate secret
* revoke secret

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8050

## How was this patch tested?
Tested manually

adoroszlai commented:
Thanks @ivanzlenko for the patch.

Please take a look at failed checks:

Checkstyle:
https://github.com/ivanzlenko/ozone/actions/runs/4598937885/jobs/8123679899#step:6:527

```
hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/S3GatewayHttpServer.java
 54: Line is longer than 80 characters (found 92).
 58: Line is longer than 80 characters (found 118).
 61: Line is longer than 80 characters (found 99).
 63: Line is longer than 80 characters (found 87).
 69: Line is longer than 80 characters (found 91).
 71: Line is longer than 80 characters (found 100).
hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/package-info.java
 1: File does not end with a newline.
hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretEndpointBase.java
 21: Unused import - org.apache.hadoop.hdds.conf.OzoneConfiguration.
 28: Unused import - org.apache.hadoop.security.HadoopKerberosName.
 33: Unused import - java.io.IOException.
hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/ServletElementsFactory.java
 7: Missing a Javadoc comment.
 9: Line is longer than 80 characters (found 96).
 20: Line is longer than 80 characters (found 112).
hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/BaseHttpServer.java
 261: First sentence should end with a period.
 268: Line is longer than 80 characters (found 91).
 268: 'name' hides a field.
```


Findbugs:
https://github.com/ivanzlenko/ozone/actions/runs/4598937885/jobs/8123680269#step:6:2286

```
M P UrF: Unread field: org.apache.hadoop.hdds.server.http.HttpServer2$Builder.keytabConfKey  At HttpServer2.java:[line 356]
M P UrF: Unread field: org.apache.hadoop.hdds.server.http.HttpServer2$Builder.usernameConfKey  At HttpServer2.java:[line 351]
```

Also, license header is needed in all files.  Example:

https://github.com/apache/ozone/blob/050d58b6d0f59e251519a4108930fa78f850b231/hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/scm/net/TestNetUtils.java#L1-L17

Rat check shows it's missing from `ServletElementsFactory`:
https://github.com/ivanzlenko/ozone/actions/runs/4598937885/jobs/8123680406#step:6:815

ivanzlenko commented:
Thanks @adoroszlai ! I've addressed all checkstyle and findbug issues.

adoroszlai reviewed (changes requested):
Please add tests for the new functionality (Robot tests in the `hadoop-ozone/dist/src/main/smoketest/s3` directory, unit tests in the usual location).

ivanzlenko commented:
@adoroszlai thanks for comments. I've uploaded requested changes.

ashishkumar50 commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretEnabledEndpointRequestFilter.java:
```diff
@@ -0,0 +1,56 @@
+    boolean isSecretEnabled = ozoneConfiguration.getBoolean(
+        OZONE_S3G_SECRET_HTTP_ENABLED_KEY, false);
+    if (!isSecretEnabled) {
+      requestContext.abortWith(Response.status(Response.Status.CONFLICT)
```
Instead of CONFLICT better to use Response.Status.BAD_REQUEST.

ashishkumar50 reviewed (changes requested):
@ivanzlenko, Thanks for the patch. Please find the comments.

ivanzlenko commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/S3GatewayHttpServer.java:
```diff
@@ -18,29 +18,88 @@
+    if (conf.getBoolean(OZONE_S3G_SECRET_HTTP_ENABLED_KEY, false)) {
+      String authType = conf.get(OZONE_S3G_SECRET_HTTP_AUTH_TYPE, "simple");
+
+      if (UserGroupInformation.isSecurityEnabled()
```
It's actually the main idea to throw exception in that case. This endpoint should be only enabled if global security is enabled.  If this endpoint is enabled without enabled security it should be treated as a missconfiguration.

ivanzlenko commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretGenerateEndpoint.java:
```diff
@@ -0,0 +1,48 @@
+  public Response get() throws IOException {
+    S3SecretResponse s3SecretResponse = new S3SecretResponse();
+    S3SecretValue s3SecretValue = generateS3Secret();
+    s3SecretResponse.setAwsSecret(s3SecretValue.getAwsSecret());
```
Wouldn't it be better to mark this interface as @NotNull?

ashishkumar50 replied:
Ya you can, in any case it should be safe enough.

ivanzlenko commented on hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3secret/TestSecretRevoke.java:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  @Test
+  void testSecretRevoke() throws IOException {
```
Can't come up with any negative test cases unfortuanatelly. Calling revoke with no secret will result in correct execution of revoke method hence this case will be indistinguishable from case with secret.

ivanzlenko commented:
@ashishkumar50 thanks, I've addressed comments, please you can take a look?

ashishkumar50 commented:
> @ashishkumar50 can you have a look at changes please?

@ivanzlenko , Thanks for updating the PR, change LGTM.

ivanzlenko commented:
@ashishkumar50 I've fixed failed test so it ready to be re-reviewed

adoroszlai commented:
@ChenSammi @Tejaskriya please take a look

ashishkumar50 commented:
> @ashishkumar50 I've fixed failed test so it's ready to be re-reviewed

@ivanzlenko,Thanks for fixing it. 
@ChenSammi Can you please help to take a look.

ivanzlenko commented:
@ChenSammi can you please check this review?

ivanzlenko commented:
@ChenSammi can you review this pull request please?

ivanzlenko commented:
@ashishkumar50 @adoroszlai can we have someone to review this ticket please?

ivanzlenko commented:
@ChenSammi thanks for review! I've addressed your comments. "globalAuth" option has been removed - it is a remnant of the first implementation and has no use in the current one.

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretConfigKeys.java:
```diff
@@ -0,0 +1,44 @@
+  public static final String OZONE_S3G_SECRET_KEYTAB_FILE =
+      OZONE_S3G_SECRET_HTTP_AUTH_CONFIG_PREFIX + "kerberos.keytab";
+  public static final String
+      OZONE_S3G_SECRET_WEB_AUTHENTICATION_KERBEROS_PRINCIPAL =
```
OZONE_S3G_SECRET_WEB_AUTHENTICATION_KERBEROS_PRINCIPAL -> 

OZONE_S3G_SECRET_HTTP_KERBEROS_PRINCIPAL

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretEndpointBase.java:
```diff
@@ -0,0 +1,99 @@
+  private OzoneClient client;
+
+  protected static final AuditLogger AUDIT =
+          new AuditLogger(AuditLoggerType.S3GLOGGER);
```
Please correct the indention.

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/S3GatewayHttpServer.java:
```diff
@@ -17,30 +17,90 @@
+  private void addSecretAuthentication(MutableConfigurationSource conf)
+      throws IOException {
+
+    if (conf.getBoolean(OZONE_S3G_SECRET_HTTP_ENABLED_KEY, false)) {
```
Please define a constant variable for this default "false" value.

ChenSammi replied:
I prefer use the constant variable. If you find any existing code which violates this code style, it's up to you to refactor it or not. But for the newly added code, I would suggest we follow the best practice coding style.

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretConfigKeys.java:
```diff
@@ -0,0 +1,43 @@
+      "ozone.s3g.secret.http.auth.";
+  public static final String OZONE_S3G_SECRET_HTTP_AUTH_TYPE =
+      OZONE_S3G_SECRET_HTTP_AUTH_CONFIG_PREFIX + "type";
+  public static final String OZONE_S3G_SECRET_KEYTAB_FILE =
```
OZONE_S3G_SECRET_HTTP_AUTH_TYPE -> 
OZONE_S3G_SECRET_HTTP_KERBEROS_KEYTAB_FILE_KEY

ChenSammi replied:
I suggest to change the variable name, to be consist with other similar variables' naming convention.

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretConfigKeys.java:
```diff
@@ -0,0 +1,43 @@
+  public static final String OZONE_S3G_SECRET_KEYTAB_FILE =
+      OZONE_S3G_SECRET_HTTP_AUTH_CONFIG_PREFIX + "kerberos.keytab";
+  public static final String OZONE_S3G_SECRET_HTTP_KERBEROS_PRINCIPAL =
+      OZONE_S3G_SECRET_HTTP_AUTH_CONFIG_PREFIX + "kerberos.principal";
```
Please add all these configuration properties in ozone-default.xml. 

There is a SecuringS3.md document about how to manage the s3 secret. Kindly update the document to add one section to this solution.

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretEndpointBase.java:
```diff
@@ -0,0 +1,99 @@
+  protected static final AuditLogger AUDIT =
+      new AuditLogger(AuditLoggerType.S3GLOGGER);
+
+  protected String shortNameFromRequest() {
```
shortNameFromRequest -> userNameFromRequest

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretRevokeEndpoint.java:
```diff
@@ -0,0 +1,47 @@
+
+  @GET
+  public Response get() throws IOException {
+    revokeSecret();
```
Is there any case that revokeSecret will fail?

ChenSammi replied:
Call revoke twice will fail with 


bash-4.2$ curl --negotiate -u : http://s3g:9878/secret/revoke
<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 500 Request failed.</title>
</head>
<body><h2>HTTP ERROR 500 Request failed.</h2>
<table>
<tr><th>URI:</th><td>/secret/revoke</td></tr>
<tr><th>STATUS:</th><td>500</td></tr>
<tr><th>MESSAGE:</th><td>Request failed.</td></tr>
<tr><th>SERVLET:</th><td>jaxrs</td></tr>
</table>

</body>
</html>

ChenSammi commented on hadoop-hdds/common/src/main/resources/ozone-default.xml:
```diff
@@ -1797,6 +1797,40 @@
+  </property>
+  <property>
+    <name>ozone.s3g.secret.http.auth.kerberos.principal</name>
+    <value>HTTP/_HOST@REALM</value>
```
We need a different default value for this property because it's the same as "ozone.s3g.http.auth.kerberos.principal".

ChenSammi commented on hadoop-hdds/common/src/main/resources/ozone-default.xml:
```diff
@@ -1797,6 +1797,40 @@
+  </property>
+  <property>
+    <name>ozone.s3g.secret.http.auth.type</name>
+    <value>simple</value>
```
I think we should set this default to "kerberos", not "simple", for simple is not allowed in the implementation.

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretEnabled.java:
```diff
@@ -0,0 +1,34 @@
+@NameBinding
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE, ElementType.METHOD})
+public @interface S3SecretEnabled {
```
Would you explain this annotation interface a bit more? I don't quite get how it enable or disable the Secure Endpoints.

ChenSammi replied:
@ivanzlenko , thanks for the info.

ChenSammi commented:
@ivanzlenko , thanks for updating the patch. Besides the inline comments, I have one more question. 
Current S3 secret CLI supports get/set/revoke secrets for the user who executing the CLI, or for other user's(with additional parameters). 
For now, the set operation and the ability to operate for other user's are supported in this patch. Do you have the plan to provide the same capability as CLI in http interface too,  probably in a follow up JIRA?

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretGenerateEndpoint.java:
```diff
@@ -0,0 +1,49 @@
+@Path("/secret/generate")
+@S3SecretEnabled
+public class S3SecretGenerateEndpoint extends S3SecretEndpointBase {
+  @GET
```
Since it will generate the secret and change server's state,  I think "@POST" is more appropriate thant "@GET, also the function name get(). 

The same suggestion apply to S3SecretRevokeEndpoint.

ChenSammi commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretConfigKeys.java:
```diff
@@ -0,0 +1,43 @@
+  public static final String OZONE_S3G_SECRET_HTTP_AUTH_TYPE_KEY =
+      OZONE_S3G_SECRET_HTTP_AUTH_CONFIG_PREFIX + "type";
+  public static final String OZONE_S3G_SECRET_HTTP_AUTH_TYPE_DEFAULT =
+      "simple";
```
simple -> kerberos

ChenSammi replied:
The property has default value in ozone-default.xml.  The same default value is required here, otherwise, it's not consistent.

ChenSammi commented on hadoop-ozone/dist/src/main/compose/ozonesecure/docker-config:
```diff
@@ -104,6 +107,8 @@ OZONE-SITE.XML_hdds.datanode.http.auth.kerberos.principal=HTTP/dn@EXAMPLE.COM
 OZONE-SITE.XML_hdds.datanode.http.auth.kerberos.keytab=/etc/security/keytabs/dn.keytab
 OZONE-SITE.XML_ozone.s3g.http.auth.kerberos.keytab=/etc/security/keytabs/s3g.keytab
 OZONE-SITE.XML_ozone.s3g.http.auth.kerberos.principal=HTTP/s3g@EXAMPLE.COM
+OZONE-SITE.XML_ozone.s3g.secret.http.auth.kerberos.keytab=/etc/security/keytabs/s3g.keytab
```
There two lines can be removed.

ChenSammi commented on hadoop-ozone/dist/src/main/compose/ozonesecure-mr/docker-config:
```diff
@@ -46,6 +50,8 @@ OZONE-SITE.XML_hdds.datanode.http.auth.kerberos.principal=HTTP/dn@EXAMPLE.COM
 OZONE-SITE.XML_hdds.datanode.http.auth.kerberos.keytab=/etc/security/keytabs/HTTP.keytab
 OZONE-SITE.XML_ozone.s3g.http.auth.kerberos.keytab=/etc/security/keytabs/s3g.keytab
 OZONE-SITE.XML_ozone.s3g.http.auth.kerberos.principal=HTTP/s3g@EXAMPLE.COM
+OZONE-SITE.XML_ozone.s3g.secret.http.auth.kerberos.keytab=/etc/security/keytabs/s3g.keytab
```
There two lines can be removed.

ChenSammi commented:
@ivanzlenko , the last patch overall looks good, just a few minor comments.

ChenSammi commented:
Thanks @ivanzlenko for the contribution.
